cmd/cstor-pool-mgmt/controller/replica-controller/handler.go

```diff
@@ -522,3 +544,98 @@ func (c *CStorVolumeReplicaController) syncCvr(cvr *apis.CStorVolumeReplica) {
 		cvr.Status.Capacity = *capacity
 	}
 }
+
+func (c *CStorVolumeReplicaController) upgrade(cvr *apis.CStorVolumeReplica) (
+	*apis.CStorVolumeReplica, error,
+) {
+	var err error
+	if cvr.VersionDetails.Current != cvr.VersionDetails.Desired {
+		if !isCurrentVersionValid(cvr) {
+			return nil, pkg_errors.Errorf("invalid current version %s", cvr.VersionDetails.Current)
+		}
+		if !isDesiredVersionValid(cvr) {
+			return nil, pkg_errors.Errorf("invalid desired version %s", cvr.VersionDetails.Desired)
+		}
+		path := strings.Split(cvr.VersionDetails.Current, "-")[0] + "-" +
+			strings.Split(cvr.VersionDetails.Desired, "-")[0]
+		u := &upgradeParams{
+			cvr:    cvr,
+			client: c.clientset,
+		}
+		cvr, err = upgradeMap[path](u)
+		if err != nil {
+			return nil, err
+		}
+		cvr.VersionDetails.Current = cvr.VersionDetails.Desired
+		cvr, err = c.clientset.OpenebsV1alpha1().
+			CStorVolumeReplicas(cvr.Namespace).Update(cvr)
+		if err != nil {
+			return nil, err
+		}
+		return cvr, nil
+	}
+	return cvr, nil
+}
+
+// populateVersion assigns VersionDetails for old cvr object
+func (c *CStorVolumeReplicaController) populateVersion(cvr *apis.CStorVolumeReplica) (
+	*apis.CStorVolumeReplica, error,
+) {
+	v := cvr.Labels[string(apis.OpenEBSVersionKey)]
+	// 1.3.0 onwards new CSP will have the field populated during creation
+	if v < "1.3.0" && cvr.VersionDetails.Current == "" {
+		cvr.VersionDetails.Current = v
+		cvr.VersionDetails.Desired = v
+		obj, err := c.clientset.OpenebsV1alpha1().CStorVolumeReplicas(cvr.Namespace).
+			Update(cvr)
+
+		if err != nil {
+			return nil, err
+		}
+		klog.Infof("Version %s added on cvr %s", v, cvr.Name)
+		return obj, nil
+	}
+	return cvr, nil
+}
+
+func isCurrentVersionValid(cvr *apis.CStorVolumeReplica) bool {
+	validVersions := []string{"1.0.0", "1.1.0", "1.2.0"}
+	version := strings.Split(cvr.VersionDetails.Current, "-")[0]
+	return util.ContainsString(validVersions, version)
+}
+
+func isDesiredVersionValid(cvr *apis.CStorVolumeReplica) bool {
+	validVersions := []string{"1.3.0"}
+	version := strings.Split(cvr.VersionDetails.Desired, "-")[0]
+	return util.ContainsString(validVersions, version)
+}
+
+type upgradeParams struct {
+	cvr    *apis.CStorVolumeReplica
+	client clientset.Interface
+}
+
+type upgradeFunc func(u *upgradeParams) (*apis.CStorVolumeReplica, error)
+
+func setReplicaID(u *upgradeParams) (*apis.CStorVolumeReplica, error) {
+	cvr := u.cvr
+	replicaID, err := cvrv1alpha1.BuilderForAPIObject(cvr).Cvr.GetReplicaID()
+	if err != nil {
+		return nil, err
+	}
+	cvr.Spec.ReplicaID = replicaID
+	cvr, err = u.client.OpenebsV1alpha1().
```

Comment:
multiple update is done on cvr because one update is done in the `setReplicaID` function and if this function does not returns any error then only in the `upgrade` function we set current version as desired version